### PR TITLE
Cache 'dn' attribute name and missing attribute names

### PR DESCRIPTION
**Pull request by rcritten**
By default an LDAP caching layer is enabled on all LDAPClient connections. The intent is to reduce the number of duplicate LDAP reads. The cache is not particularly aggressive in storing data so there is room for improvement.

A cache MISS occurs when only the 'dn' attribute is requested because while this attribute is available in all cache entries the 'dn' name does not appear in the attribute list.

Missing attribute names are not cached. This is seen most often in the cn=ipaconfig request. By default there are no values for 'ipauserdefaultsubordinateid', 'ipadomainresolutionorder' and 'ipauserauthtype'. So since those are always requested and not returned the content of cn=ipaconfig is always treated as a cache MISS.

The LDAP connections are generally short-lived so I don't think we have too much risk of making decisions based on non-cached values.

These two changes raise the cache hit ratio from 35 - 43% for ipa-healthcheck.

Fixes: https://pagure.io/freeipa/issue/9475

### CODE REVIEW

**Comment by rcritten** [comment]
One of the xmlrpc tests is failing:

0021: service_add_host: Add mixed-case host to 'HTTP/testhost1.example.test@EXAMPLE.TEST'

E           AssertionError: assert_deepequal: expected != got.
E             0021: service_add_host: Add mixed-case host to 'HTTP/testhost1.example.test@EXAMPLE.TEST'
E             expected = 'testhost1.example.test'
E             got = 'TestHost3.example.test'
E             path = ('result', 'managedby_host', 0)

File: ipapython/ipaldap.py [hunk 1]
```diff
@@ -1998,6 +1999,8 @@ def get_entry(self, dn, attrs_list=None, time_limit=None,
         if entry and attrs_list:
             req_attrs = set(attr.lower() for attr in set(attrs_list))
             cache_attrs = set(attr.lower() for attr in entry.attrs_list)
+            # dn is stored separately but is still available as an attr
+            cache_attrs.update({'dn'})
```
Comment (tiran) [inline ipapython/ipaldap.py:2003]:
```suggestion
            cache_attrs.add('dn')
```

**Comment by rcritten** [comment]
Something about these changes is modifying the rule in a way that sssd doesn't like. I've duplicated it manually. But comparing a working a non-working entry both in the IPA tree and ou=sudoers they appear to be identical (beyond ipaUniqueId). SSSD shouldn't be affected by the cache since it is only on the API level.

**Comment by rcritten** [comment]
SSSD is case-sensitive to member DN values. It has to be the form of ipaUniqueID. All lower case is invalid. I'll open a bug.

**Comment by rcritten** [comment]
This is caused by the change: dn = DN(str(dn).lower())

I'll resubmit without this optimization since this is going to break SSSD all the way back to the beginning most likely.

**Comment by rcritten** [comment]
Apparently SSSD only works with camel-case member values (e.g. ipaUniqueId). I filed an upstream ticket.
I'm going to close this for now. Given that SSSD has done this since time and memorial any changes to casing is going to break one client or another. So I'm going to abandon this change for now.

**Pull request closed without merging**